spec/index.adoc

```diff
@@ -3996,9 +4016,74 @@ S with
+
+We define an auxiliary method that handles calls from composite query methods by performing a call graph traversal. It can also be (trivially) invoked for query methods that do not make further calls.
+....
+composite_query_helper(S, Cycles, Caller, Canister_id, Method_name, Arg) =
+  let Mod = S.canisters[Canister_id].module
+  let F = Mod.composite_query_methods[Method_name]
+  let Cert <- { Cert | verify_cert(Cert) and
+                       lookup(["canister", Canister_id, "certified_data"], Cert) = Found S.certified_data[Canister_id] and
+                       lookup(["time"], Cert) = Found S.system_time // or “recent enough”
+              }
+  let Env = { time = S.time[Canister_id];
+              global_timer = S.global_timer[Canister_id];
+              balance = S.balances[Canister_id];
+              freezing_limit = freezing_limit(S, Canister_id);
+              certificate = Cert;
+              status = simple_status(S.canister_status[Canister_id]);
+              canister_version = S.canister_version[Canister_id];
+            }
+  if S.canisters[Canister_id] ≠ EmptyCanister and
+     S.canister_status[Canister_id] = Running and
+     S.balances[Canister_id] >= freezing_limit(S, Canister_id) and
+     Method_name ∈ dom(Mod.composite_query_methods) and
+     Cycles >= MAX_CYCLES_PER_MESSAGE
+  then
+     let W = S.canisters[Canister_id].wasm_state
+     let R = F(Arg, Caller, Env)(W)
+     if R = Trap trap
+     then Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - trap.cycles_used)
+     else if R = Return {new_state = W'; new_calls = Calls; response = Response; cycles_used = Cycles_used}
+     then
+        W := W'
+        Cycles := Cycles - Cycles_used
+        if Response = NoResponse
+        then
+           while Calls ≠ []
+           do
+              let Calls' · Call · Calls''  = Calls
+              Calls := Calls' · Calls''
+              if S.canister_subnet[Canister_id] ≠ S.canister_subnet[Call.callee]
+              then
+                 Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // calling to another subnet
+              let (Response', Cycles_used') = composite_query_helper(S, Cycles, Canister_id, Call.callee, Call.method_name, Call.arg)
+              Cycles := Cycles - Cycles_used'
+              if Cycles < MAX_CYCLES_PER_RESPONSE
+              then
+                 Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // composite query out of cycles
+              let F' = Mod.callbacks(Call.callback, Response', 0, Env, 0)
+              let R'' = F'(W')
+              if R'' = Trap trap''
+              then Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - trap''.cycles_used)
+              else if R'' = Return {new_state = W''; new_calls = Calls''; response = Response''; cycles_used = Cycles_used''}
+              then
+                 W := W''
```

Comment:
We don't commit `W''` here.

Reply:
This change is not committed, but necessary to process callbacks properly. Otherwise, the callbacks would run on the original canister state and not see changes made before `.await`.